src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,375 @@
+        classifier._set_override_scaling_pass_value(value)
+
+    def get_global_scaling_weight(self) -> float:
+        """
+        Get the global LoRA weight.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        return classifier.config.global_scaling_weight
+
+    def get_latest_scalings(self) -> Optional[torch.Tensor]:
+        """
+        Returns the latest scalings prediction, or None if no scalings have been predicted. The tensor is of shape
+        (batch_size, seq_len, n_layers, n_classes).
+        """
+        return self.internal_xlora_scalings
+
+    def get_scalings_log(self) -> List[torch.Tensor]:
+        """
+        Returns a shallow (only copying the list itself not the tensors) copy of the list containing the scalings log.
+        Editing the list does not change the underlying log. The tensors are of shape (batch_size, seq_len, n_layers,
+        n_classes). The seq_len dim may vary with input dimension.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        return classifier.log_scalings.copy()
+
+    def enable_scalings_logging(self):
+        """
+        Enable scalings logging.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        classifier.scalings_logging = True
+
+    def disable_scalings_logging(self):
+        """
+        Disable scalings logging, without clearing the log.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        classifier.scalings_logging = False
+
+    def clear_scalings_log(self):
+        """
+        Clear the scalings log.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        classifier.log_scalings.clear()
+
+    def get_bucketed_scalings_log(self) -> dict[int, tuple[list[int], list[torch.Tensor]]]:
+        """
+        Returns bucketed scalings, bucketed by seq_len. Each value consists of the positions (the first) and the
+        associated tensors. The positions are paired with the associated tensors and give the position in the scaling
+        log.
+        """
+        classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+        return classifier._get_bucketed_scalings()
+
+    def set_use_trainable_adapters(self, use_trainable_adapters: bool):
+        """
+        Set the adapters to trainable or not trainable.
+
+        This is reflected in the config.
+        """
+        for name, param in self.named_parameters():
+            if "lora_" in name:
+                param.requires_grad = use_trainable_adapters
+
+        self.xlora_config.use_trainable_adapters = use_trainable_adapters
+
+    def get_use_trainable_adapters(self) -> bool:
+        """
+        Get the trainable or not trainable state of the adapters.
+        """
+        return self.xlora_config.use_trainable_adapters
```

Comment:
Not really sure if we need these two method. We should set this through the config, unless there is a reason why we need to dynamically change the value, which I don't see right now.